### PR TITLE
Stories/dt/keyword density

### PR DESCRIPTION
**Pull request by terw-dan**
created keyword density function and subheading function. Updated function queuer and added comments to function.

### CODE REVIEW

File: js/analyzer.js [hunk 1]
```diff
@@ -1,65 +1,232 @@
...
+};
 
-/*text preprocessor*/
+/**
+ * Analyzer initialisation. Loads defaults and overloads custom settings.
+ */
+Analyzer.prototype.init = function() {
 
+    //init preprocessor if not exists
+    if (typeof yst_pp != 'object' || yst_pp.inputText != this.config.textString) {
```
Comment (CarolineGeven) [inline js/analyzer.js]: Please do not abbreviate `pp`. I assume this means preprocessor?

File: js/analyzer.js [hunk 2]
```diff
@@ -1,65 +1,232 @@
...
+    return yst_pp.stripSpaces(textString);
+};
+
+/**
+ * PreProcessor object definition. Creates _store object and calls init.
+ */
 PreProcessor = function (text){
     //create _store object to store data
     this._store = {};
     this._store.origText = text;
```
Comment (CarolineGeven) [inline js/analyzer.js]: Please don't abbreviate `origText`

Comment (CarolineGeven) [inline js/analyzer.js]: Please don't abbreviate `origText`

File: js/analyzer.js [hunk 3]
```diff
@@ -1,65 +1,232 @@
...
+    //remove first function from queue and execute it.
+    if(this.queue.length > 0){
+        this._output.push(this[this.queue.shift()]());
+        this.runQueue();
+    }
+};
+/**
+ * clears current queue of functions, effectively stopping execution of the analyzer.
+ */
+Analyzer.prototype.abortQueue = function(){
```
Comment (CarolineGeven) [inline js/analyzer.js]: In which cases is this function used?

File: js/analyzer.js [hunk 4]
```diff
@@ -1,65 +1,232 @@
...
+ * @returns resultObject
+ */
+Analyzer.prototype.keywordDensity = function(){
+    if(yst_pp._store.wordcount >= 100) {
+        var keywordMatches = yst_pp._store.cleanText.match(new RegExp(this.config.keyword, 'g'));
+        if ( keywordMatches != null ) {
+            var keywordCount = keywordMatches.length;
+            var keywordDensity = (keywordCount / yst_pp._store.wordcount - (keywordCount - 1 * keywordCount)) * 100;
+        } else {
+            var keywordDensity = 0;
```
Comment (CarolineGeven) [inline js/analyzer.js]: For readability I would advise to move keywordDensity = 0 above the if statement. This way you don't need the else statement and is it more clear the switch will not be called if keywordDensity is 0

File: js/analyzer.js [hunk 5]
```diff
@@ -1,65 +1,232 @@
...
+    this.queue = [];
+};
+
+
+/**
+ * checks the keyword density of given keyword against the cleantext stored in _store.
+ * @returns resultObject
+ */
+Analyzer.prototype.keywordDensity = function(){
+    if(yst_pp._store.wordcount >= 100) {
```
Comment (CarolineGeven) [inline js/analyzer.js]: In the original PHP code the statement checks if > 100. Not >= 100

File: js/analyzer.js [hunk 6]
```diff
@@ -1,65 +1,232 @@
...
+            if (formattedHeaders.match(new RegExp(this.config.keyword, 'g')) || headers[i].match(new RegExp(this.config.keyword, 'g'))) {
+                foundInHeader++;
+            }
+        }
+        if (foundInHeader > 0) {
+            var result = {keywordFound: foundInHeader, numberofHeaders: headers.length};
+            var rating = 9;
+        }else{
+            var result = {keywordFound: 0, numberofHeaders: headers.length}
+            var rating = 7;
```
Comment (CarolineGeven) [inline js/analyzer.js]: According to the PHP code the rating should be 3 (https://github.com/Yoast/wordpress-seo/blob/trunk/admin/class-metabox.php#L1952)

File: js/analyzer.js [hunk 7]
```diff
@@ -1,65 +1,232 @@
...
+Analyzer.prototype.keywordDensity = function(){
+    if(yst_pp._store.wordcount >= 100) {
+        var keywordMatches = yst_pp._store.cleanText.match(new RegExp(this.config.keyword, 'g'));
+        if ( keywordMatches != null ) {
+            var keywordCount = keywordMatches.length;
+            var keywordDensity = (keywordCount / yst_pp._store.wordcount - (keywordCount - 1 * keywordCount)) * 100;
+        } else {
+            var keywordDensity = 0;
+        }
+        switch (true) {
```
Comment (CarolineGeven) [inline js/analyzer.js]: Please move the rating to a different function, so the functions have a single responsibility. In the case of `keywordDensity` I would expect a function that gets the keyword density. The keyword density rating would be a different function.

File: js/analyzer.js [hunk 8]
```diff
@@ -1,65 +1,232 @@
...
+        result = {name: 'subHeadings', result: null, rating: 7};
+    }else {
+        var foundInHeader = 0;
+        for (var i = 0; i < headers.length; i++) {
+            var formattedHeaders = this.stripKeywords(headers[i]);
+            if (formattedHeaders.match(new RegExp(this.config.keyword, 'g')) || headers[i].match(new RegExp(this.config.keyword, 'g'))) {
+                foundInHeader++;
+            }
+        }
+        if (foundInHeader > 0) {
```
Comment (CarolineGeven) [inline js/analyzer.js]: Please move the rating to a different function, so the functions have a single responsibility. In the case of subhheaderChecker I would expect a function that checks the subheader. The subheader rating would be a different function.

File: js/analyzer.js [hunk 9]
```diff
@@ -1,65 +1,232 @@
...
+        result = {name: 'subHeadings', result: result, rating: rating};
+    }
+    return result;
+};
+
+/**helper functions*/
+
+/**
+ * removes certain words from string
+ * @params textString and
```
Comment (CarolineGeven) [inline js/analyzer.js]: I'm missing the `@returns` in the docs.

File: js/analyzer.js [hunk 10]
```diff
@@ -1,65 +1,232 @@
...
+        wordString += '('+this.config.wordsToRemove[i]+')\\b';
+    }
+    var wordsRegex = new RegExp(wordString, 'g');
+    textString =  textString.replace(wordsRegex, '');
+    //remove double space
+    return yst_pp.stripSpaces(textString);
+};
+
+/**
+ * PreProcessor object definition. Creates _store object and calls init.
```
Comment (CarolineGeven) [inline js/analyzer.js]: 'text' variable is missing in the docs

File: js/analyzer.js (hunk 1 above)
Comment (andizer) [inline js/analyzer.js]: Can we do strict comparisons (probably CodeClimate will yell about this in the future) ?

File: js/analyzer.js [hunk 11]
```diff
@@ -1,65 +1,232 @@
...
+ * Analyzer initialisation. Loads defaults and overloads custom settings.
+ */
+Analyzer.prototype.init = function() {
 
+    //init preprocessor if not exists
+    if (typeof yst_pp != 'object' || yst_pp.inputText != this.config.textString) {
+        yst_pp = new PreProcessor(this.config.textString);
+    }
+    //if no available function queues, make new queue
+    if (typeof this.queue != 'Array') {
```
Comment (andizer) [inline js/analyzer.js]: I would make a function of line 20-29 to set the queue, because it's a single responsibility

**Comment by andizer** [comment]
CR done :+1:

File: js/analyzer.js (hunk 3 above)
Comment (terw-dan) [inline js/analyzer.js]: Isn't used yet. For future testing and to be able to interrupt the process.

**Pull request merged**